src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,178 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Malte Grave, LIT CPS Johannes Kepler Universität,
+ *                                 github.com/gravemalte
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Malte Grave - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+
+#include <cerrno>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <fcntl.h>
+#include <linux/can.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <vector>
+
+#include "CANLayer.h"
+#include "basecommfb.h"
+#include "comtypes.h"
+#include "datatype.h"
+#include "devlog.h"
+#include "forte_any.h"
+#include "forte_array.h"
+
+using namespace forte::com_infra;
+
+CCANComLayer::CCANComLayer(CComLayer *paUpperLayer, CBaseCommFB *paComFB)
+    : CComLayer(paUpperLayer, paComFB) {}
+
+EComResponse CCANComLayer::openConnection(char *paLayerParameter) {
+  EComResponse eRetVal = e_InitInvalidId;
+  CParameterParser parser(paLayerParameter, ';', eCANComParamterAmount);
+  if (eCANComParamterAmount != parser.parseParameters()) {
+    DEVLOG_ERROR("[CAN Layer] Parsing of the parameters failed.\n");
+    return eRetVal;
+  }
+
+  SCANParameters can_params = setupCANInternals(parser);
+
+  switch (mFb->getComServiceType()) {
+  case e_Server:
+    DEVLOG_ERROR("[CAN Layer] the layer only supports the PUB/SUB "
+                 "pattern. Please use a Publisher block to send data.");
+    eRetVal = e_InitTerminated;
+    break;
+  case e_Client:
+    DEVLOG_ERROR("[CAN Layer] the layer only supports the PUB/SUB "
+                 "pattern. Please use a Subscribe block to receive data.");
+    eRetVal = e_InitTerminated;
+    break;
+  case e_Publisher:
+    // is handled via sendData
+    break;
+  case e_Subscriber:
+    // is handled via recvData
+    break;
+  }
+
+  DEVLOG_DEBUG("[CAN Layer] Opening connection\n");
+
+  CFDSelectHandler::TFileDescriptor fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+  if (CFDSelectHandler::scmInvalidFileDescriptor == fd) {
+    DEVLOG_ERROR("[CAN Layer] Socket can't be created: %s\n", strerror(errno));
+    return e_InitInvalidId;
+  }
+
+  mCANSocket = fd;
```

Comment:
this should go right at the end, only when you're sure that the socket is properly set. 

probably next to `mFrame.can_id = can_params.CANId;`